Add plugin settings metadata to plugin info V3 (#3599)

ketan opened the pull request.

varshavaradarajan commented on plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/packagematerial/PackageMaterialPluginInfoBuilder.java:
```diff
@@ -0,0 +1,84 @@
+        if (repositoryConfiguration == null) {
+            throw new RuntimeException(format("Plugin[%s] returned null repository configuration", descriptor.id()));
+        }
+        if (packageConfiguration == null) {
```
Don't we have to do something similar for elastic profile? And probably others. There is no need to throw an exception when the plugin settings is null, but when the profile settings, or scm settings, or task settings is null, we should throw an exception, right?

ketan replied:
Ideally, when we do throw up, the server should continue with the other plugins, I think we're likely failing on the first failure. Do you mind verifying?

varshavaradarajan replied:
It continues. I threw an Excpetion in pluginInfosFor method in PackageMaterialPluginInfoBuilder and ElasticAgentPluginInfoBuilder and started the server with a EA plugin and the package repo plugin, the server logs show both the exceptions. Is that a good enough verification?

varshavaradarajan commented:
Pending work - 
1. Currently, if the plugin is not loaded, the plugins page (go/admin/plugins) shows this up in grey, with a message as to why the plugin could not be loaded. The plugin_infos API needs to reflect that. 
The plugin infos index API should include the installed but not loaded plugin. The plugin_infos show call should provide the vendor details about such a plugin and the reason why it could not be loaded.

maheshp commented:
Listing an unavailable plugin as part of the API seems not ok, IMO. Instead can we show it in Errors and Warnings, I know this would be moving away from the current behaviour.

varshavaradarajan commented:
@maheshp , why is it not okay? As long as we provide the reason as to why the plugin can't be loaded..

ketan commented:
> Listing an unavailable plugin as part of the API seems not ok, IMO. Instead can we show it in Errors and Warnings, I know this would be moving away from the current behaviour.

The current behavior of the plugins page is to show the plugin, with errors on it, which is why we're recording errors in the plugin info object (and therefore the API) to show it on the upcoming plugins SPA instead of the obscure errors and warnings dialog.

ketan commented:
Fat fingered the send button. Do you think there's a better way to list the bad plugins along with the errors on the plugins page?

maheshp commented:
I would have preferred the PluginInfo API to list only the available plugins, guess you guys are right for this use case we might have to list the plugins with errors. This means we might have to introduce state to the plugin, something below makes sense?
```
{
  "id": "my_plugin",
  "name": "My SCM Plugin",
  "state": "loaded"
  ....
}
```

```
{
"id": "your_plugin",
  "name": "Your SCM Plugin",
  "state": "error"
  "errors": [
  ]	
}
```

ketan commented:
> This means we might have to introduce state to the plugin, something below makes sense?

Yes, we're thinking of adding an `error` property to the `PluginInfo` type that can help answer what the `state` of the plugin is.

maheshp commented:
I think having a `state` would help in filtering. Apart from the Plugin listing API other SPA's can just pull loaded plugins and don;t have to deal with plugins with errors.

varshavaradarajan commented:
+1 to what Mahesh's saying. No need to list the unloaded plugins on pages like elastic agents.

ketan commented on plugin-infra/go-plugin-access/test/com/thoughtworks/go/plugin/access/pluggabletask/PluggableTaskPluginInfoBuilderTest.java:
```diff
@@ -0,0 +1,104 @@
+        assertThat(pluginInfo.getDisplayName(), is("my task plugin"));
+        assertThat(pluginInfo.getTaskSettings(), is(new PluggableInstanceSettings(pluginConfigurations, pluginView)));
+    }
+}
```
That is not possible, because if the plugin does not provide a task or settings, the `JsonBasedTaskExtensionHandler_V1` will blow up appropriately.

varshavaradarajan replied:
Then remove the null checks from PluggableTaskPluginInfoBuilder.

varshavaradarajan commented:
Also, should it provide links if it is a BadPluginInfo ?

ketan commented:
> Also, should it provide links if it is a BadPluginInfo ?

You mean the `GET` call for a `BadPluginInfo`, if you can list it, you should be able to `GET` it.

varshavaradarajan commented:
The `GET` for a BadPluginInfo is 404

ketan commented:
> The GET for a BadPluginInfo is 404

Fixed

ketan commented:
I was supposed to add a note about how the API looks like:

In the response below for the `index` call, we're listing out all plugins including the ones that could not be loaded because of some issues with the plugin.

In terms of the what this PR brings to the existing API:

* `plugin_settings` a hash identical to `profile_settings` or `role_settings` object
* `status` - an object with a `state` (`active` or `disabled`) and error `messages`.
* `plugin_file_location` the location of the plugin jar
* `bundled_plugin` boolean indicating whether the plugin is bundled or not
* added support for config repo plugins
* added attributes for auth plugin which was earlier mostly blank (since it has been deprecated)

```bash
curl --request GET \
  --url 'http://localhost:8153/go/api/admin/plugin_info?include_bad=true' \
  --header 'accept: application/vnd.go.cd.v3+json' \
  --user 'admin:badger'
```
```json
{
  "_links": {
    "self": {
      "href": "http://localhost:8153/go/api/admin/plugin_info"
    },
    "find": {
      "href": "http://localhost:8153/go/api/admin/plugin_info/:plugin_id"
    },
    "doc": {
      "href": "https://api.gocd.org/#plugin-info"
    }
  },
  "_embedded": {
    "plugin_info": [
      {
        "_links": {
          "self": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/cd.go.contrib.elastic-agent.docker-swarm"
          },
          "doc": {
            "href": "https://api.gocd.org/#plugin-info"
          },
          "find": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/:plugin_id"
          },
          "image": {
            "href": "http://localhost:8153/go/api/plugin_images/cd.go.contrib.elastic-agent.docker-swarm/176f84f4804cd55e556170df3ecb88dacf40d733077e5251b4ea6048a2257d5c"
          }
        },
        "id": "cd.go.contrib.elastic-agent.docker-swarm",
        "version": "1",
        "type": "elastic-agent",
        "status": {
          "state": "active"
        },
        "plugin_file_location": "/Users/ketanpadegaonkar/projects/gocd/gocd/server/plugins/external/docker-swarm-elastic-agents-1.1.4.jar",
        "bundled_plugin": false,
        "about": {
          "name": "GoCD Docker Swarm Elastic Agents",
          "version": "1.1.4",
          "target_go_version": "16.12.0",
          "description": "Docker Swarm Based Elastic Agent Plugins for GoCD",
          "target_operating_systems": [],
          "vendor": {
            "name": "GoCD Contributors",
            "url": "https://github.com/gocd-contrib/docker-swarm-elastic-agents"
          }
        },
        "profile_settings": {
          "configurations": [
            {
              "key": "Image",
              "metadata": {
                "secure": false,
                "required": true
              }
            },
            {
              "key": "Command",
              "metadata": {
                "secure": false,
                "required": false
              }
            },
            {
              "key": "Environment",
              "metadata": {
                "secure": false,
                "required": false
              }
            },
            {
              "key": "MaxMemory",
              "metadata": {
                "secure": false,
                "required": false
              }
            },
            {
              "key": "ReservedMemory",
              "metadata": {
                "secure": false,
                "required": false
              }
            },
            {
              "key": "Hosts",
              "metadata": {
                "secure": false,
                "required": false
              }
            }
          ],
          "view": {
            "template": "<html/>"
          }
        },
        "plugin_settings": {
          "configurations": [
            {
              "key": "go_server_url",
              "metadata": {
                "secure": false,
                "required": true
              }
            },
            {
              "key": "environment_variables",
              "metadata": {
                "secure": false,
                "required": false
              }
            },
            {
              "key": "max_docker_containers",
              "metadata": {
                "secure": false,
                "required": true
              }
            },
            {
              "key": "docker_uri",
              "metadata": {
                "secure": false,
                "required": true
              }
            },
            {
              "key": "auto_register_timeout",
              "metadata": {
                "secure": false,
                "required": true
              }
            },
            {
              "key": "docker_ca_cert",
              "metadata": {
                "secure": true,
                "required": false
              }
            },
            {
              "key": "docker_client_key",
              "metadata": {
                "secure": true,
                "required": false
              }
            },
            {
              "key": "docker_client_cert",
              "metadata": {
                "secure": true,
                "required": false
              }
            }
          ],
          "view": {
            "template": "<html/>"
          }
        }
      },
      {
        "_links": {
          "self": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/cd.go.authentication.passwordfile"
          },
          "doc": {
            "href": "https://api.gocd.org/#plugin-info"
          },
          "find": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/:plugin_id"
          },
          "image": {
            "href": "http://localhost:8153/go/api/plugin_images/cd.go.authentication.passwordfile/9fdf150dbce01084106d76b18e3215af1f9c2beff6f3638e38083237f433662f"
          }
        },
        "id": "cd.go.authentication.passwordfile",
        "version": "1",
        "type": "authorization",
        "status": {
          "state": "active"
        },
        "plugin_file_location": "/Users/ketanpadegaonkar/projects/gocd/gocd/server/plugins/bundled/filebased-authentication-plugin.jar",
        "bundled_plugin": true,
        "about": {
          "name": "Password File Authentication Plugin for GoCD",
          "version": "a88e22cdef9a6415b435ed1a02d8c3727427b2ab",
          "target_go_version": "17.5.0",
          "description": "GoCD Authorization plugin for file based password authentication",
          "target_operating_systems": [],
          "vendor": {
            "name": "GoCD Contributors",
            "url": "https://github.com/gocd-contrib/filebased-authentication-plugin"
          }
        },
        "auth_config_settings": {
          "configurations": [
            {
              "key": "PasswordFilePath",
              "metadata": {
                "secure": false,
                "required": true
              }
            }
          ],
          "view": {
            "template": "<html/>"
          }
        },
        "capabilities": {
          "can_search": true,
          "supported_auth_type": "Password",
          "can_authorize": false
        }
      },
      {
        "_links": {
          "self": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/slack.notifier"
          },
          "doc": {
            "href": "https://api.gocd.org/#plugin-info"
          },
          "find": {
            "href": "http://localhost:8153/go/api/admin/plugin_info/:plugin_id"
          }
        },
        "id": "slack.notifier",
        "version": "1",
        "type": null,
        "status": {
          "state": "invalid",
          "messages": [
            "Class [GoNotificationPlugin] is annotated with @Extension but cannot be constructed. Reason: java.lang.RuntimeException: Unable to find go_notify.conf. Please make sure you've set it up right..",
            "No extensions found in this plugin.Please check for @Extension annotations"
          ]
        },
        "plugin_file_location": "/Users/ketanpadegaonkar/projects/gocd/gocd/server/plugins/external/gocd-slack-notifier-1.4.0-RC11.jar",
        "bundled_plugin": false,
        "about": {
          "name": "Slack Notification Plugin",
          "version": "1.4.0-RC11",
          "target_go_version": "16.3.0",
          "description": "Plugin to send build notifications to slack",
          "target_operating_systems": [],
          "vendor": {
            "name": "Ashwanth Kumar",
            "url": "https://github.com/ashwanthkumar/gocd-slack-build-notifier"
          }
        }
      }
    ]
  }
}
```
